xdsl/transforms/loop_hoist_memref.py

```diff
@@ -0,0 +1,144 @@
+from dataclasses import dataclass
+
+from xdsl.dialects import builtin, memref, scf
+from xdsl.ir import Block, Operation, Region
+from xdsl.irdl import Operand
+from xdsl.passes import MLContext, ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.transforms.utils import (
+    find_same_target_store,
+    get_operation_at_index,
+    is_loop_dependent,
+)
+
+
+@dataclass
+class LoopHoistMemref(RewritePattern):
+    """Hoist pairs of memref.loads and memref.stores out of a loop."""
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self,
+        for_op: scf.For,
+        rewriter: PatternRewriter,
+    ) -> None:
+        if for_op.parent_block() is None:
+            return
+
+        loads: list[memref.Load] = []
+        for op in for_op.body.ops:
+            if isinstance(op, memref.Load):
+                loads.append(op)
+
+        if len(loads) == 0:
+            return
+
+        parent_block = loads[0].parent_block()
+        assert parent_block is not None
+
+        # filter out multiple loads from the same location
+        load_locs = [load.memref for load in loads]
+        dup_load_locs = [loc for loc in set(load_locs) if load_locs.count(loc) > 1]
+        loads = [load for load in loads if load.memref not in dup_load_locs]
+
+        load_store_pairs: dict[memref.Load, memref.Store] = {}
+
+        for load in loads:
+            if (
+                (store := find_same_target_store(load))
+                and parent_block.get_operation_index(load)
+                < parent_block.get_operation_index(store)
+                and not any(is_loop_dependent(idx, for_op) for idx in load.indices)
+            ):
+                load_store_pairs[load] = store
+
+        # filter out stores using the same value
+        store_vals = [store.value for store in load_store_pairs.values()]
+        dup_store_vals = [val for val in store_vals if store_vals.count(val) > 1]
+        load_store_pairs = {
+            load: store
+            for load, store in load_store_pairs.items()
+            if store.value not in dup_store_vals
+        }
+
+        if len(load_store_pairs.items()) == 0:
+            return
+
+        # hoist new loads before the current loop
+        new_loads = [load.clone() for load in load_store_pairs.keys()]
+        rewriter.insert_op_before(new_loads, for_op)
+
+        ld_indices = [
+            parent_block.get_operation_index(load) for load in load_store_pairs.keys()
+        ]
+        st_indices = [
+            parent_block.get_operation_index(store)
+            for store in load_store_pairs.values()
+        ]
```

Comment:
why go through the indices and not the ops directly?

Reply:
I want a map between the original to cloned block ops, and not all ops, just load and stores.
Till the `clone_into` offers that, this is the simplest way to filter them once we have the targets to operate on.

Reply:
I'll see if I can change it closer to the proposed API change.